packages/firestore/test/unit/util/bundle.test.ts

```diff
@@ -0,0 +1,332 @@
+    verifySizedElement(actual[1], limitToLastQuery, limitToLastQueryString);
+    verifySizedElement(actual[2], doc1Meta, doc1MetaString);
+    verifySizedElement(actual[3], doc1, doc1String);
+  });
+
+  it(
+    'reads with unexpected orders with bytesPerRead ' + bytesPerRead,
+    async () => {
+      const bundleStream = readableStreamFromString(
+        metaString +
+          doc1MetaString +
+          doc1String +
+          limitQueryString +
+          doc2MetaString +
+          doc2String,
+        bytesPerRead
+      );
+      const bundle = new BundleReader(bundleStream);
+
+      const actual = await getAllElement(bundle);
+      expect(actual.length).to.equal(5);
+      verifySizedElement(actual[0], doc1Meta, doc1MetaString);
+      verifySizedElement(actual[1], doc1, doc1String);
+      verifySizedElement(actual[2], limitQuery, limitQueryString);
+      verifySizedElement(actual[3], doc2Meta, doc2MetaString);
+      verifySizedElement(actual[4], doc2, doc2String);
+
+      // Reading metadata after other elements should also work.
+      expect(await bundle.getMetadata()).to.deep.equal(meta.metadata);
+    }
+  );
+
+  it(
+    'reads without named query with bytesPerRead ' + bytesPerRead,
+    async () => {
+      const bundleStream = readableStreamFromString(
+        metaString + doc1MetaString + doc1String,
+        bytesPerRead
+      );
+      const bundle = new BundleReader(bundleStream);
+
+      expect(await bundle.getMetadata()).to.deep.equal(meta.metadata);
+
+      const actual = await getAllElement(bundle);
+      expect(actual.length).to.equal(2);
+      verifySizedElement(actual[0], doc1Meta, doc1MetaString);
+      verifySizedElement(actual[1], doc1, doc1String);
+    }
+  );
+
+  it('reads with deleted doc with bytesPerRead ' + bytesPerRead, async () => {
+    const bundleStream = readableStreamFromString(
+      metaString + noDocMetaString + doc1MetaString + doc1String,
+      bytesPerRead
+    );
+    const bundle = new BundleReader(bundleStream);
+
+    expect(await bundle.getMetadata()).to.deep.equal(meta.metadata);
+
+    const actual = await getAllElement(bundle);
+    expect(actual.length).to.equal(3);
+    verifySizedElement(actual[0], noDocMeta, noDocMetaString);
+    verifySizedElement(actual[1], doc1Meta, doc1MetaString);
+    verifySizedElement(actual[2], doc1, doc1String);
+  });
+
+  it(
+    'reads without documents or query with bytesPerRead ' + bytesPerRead,
+    async () => {
+      const bundleStream = readableStreamFromString(metaString, bytesPerRead);
+      const bundle = new BundleReader(bundleStream);
+
+      expect(await bundle.getMetadata()).to.deep.equal(meta.metadata);
+
+      const actual = await getAllElement(bundle);
+      expect(actual.length).to.equal(0);
+    }
+  );
+
+  it(
+    'throws with ill-formatted bundle with bytesPerRead ' + bytesPerRead,
+    async () => {
+      await expect(
+        parseThroughBundle('metadata: "no length prefix"', bytesPerRead)
```

Comment:
In the Web SDK, we generally assert on the actual error messages as this allows us to verify that the messages make sense. Do you mind doing this here? You can look at `validation.test.ts` for examples.

Reply:
Done.